# Scripts
## [Geographic Maps](https://github.com/tshalev/WRC-genome-paper/tree/main/Supplementary code/geographic_maps)
Geographic map used in Figure 1

## [Genetic Differentiation](https://github.com/tshalev/WRC-genome-paper/main/Supplementary code/genetic_differentiation)
Fst analysis, DAPC, PCA plots used in Figure 1

## [LD Plots](https://github.com/tshalev/WRC-genome-paper/tree/main/Supplementary code/LD_plots)
Generating LD plots used in Figure 1. Code derived and adapted from a post made here: https://www.biostars.org/p/347796/

## [Nucleotide Diversity Analysis](https://github.com/tshalev/WRC-genome-paper/main/Supplementary code/nucleotide_diversity_analysis)
Analysis of pi, dxy in pixy, plots used in Figure 3, statistical tests

## [S Line Genotype Correction](https://github.com/tshalev/WRC-genome-paper/main/Supplementary code/S_line_genotype_correction)
Manual correction of SNP data for complete selfing lines. Further fine-scale manual correction was carried out by hand.

## [S Line Pop Gen and Heterozygosity](https://github.com/tshalev/WRC-genome-paper/tree/main/Supplementary code/S_line_pop_gen_and_heterozygosity)
Analysis of heterozygosity, statistical tests, chi-squared test for analysis of loci under selection during selfing (outlier SNPs), plots used in Figure 4

## [Inbreeding Coeffcients](https://github.com/tshalev/WRC-genome-paper/main/Supplementary code/inbreeding_coefficients)
Analysis and plots for inbreeding coefficients in Figure 4

## [Variant Effect Predictions and SNP Annotation](https://github.com/tshalev/WRC-genome-paper/main/Supplementary code/variant_effect_predictions_and_snp_annotation)
Analysis of outlier SNPs during selfing, variant effects, statistical analyses of SNP consequences and GO categories, annotation of SNPs
