# Scripts
## [geographic_maps.R](https://github.com/tshalev/WRC-genome-paper/blob/main/Scripts/geographic_maps.R)
Geographic maps used in Figure 1

## [genetic_differentiation.R](https://github.com/tshalev/WRC-genome-paper/blob/main/Scripts/genetic_differentiation.R)
Fst analysis, DAPC, AMOVA, PCA plots used in Figure 2

### [PopGenome_analysis.R](https://github.com/tshalev/WRC-genome-paper/blob/main/Scripts/PopGenome_analysis.R)
Analysis of pi, theta, Wall's B in PopGenome, normalisation by gene length, statistical tests

### [LD_plots.R](https://github.com/tshalev/WRC-genome-paper/blob/main/Scripts/LD_plots.R)
Generating LD plots used in Figure 3. Code derived and adapted from a post made here: https://www.biostars.org/p/347796/

### [LD_pruning.R](https://github.com/tshalev/WRC-genome-paper/blob/main/Scripts/LD_pruning.R)
Pruning SNP data based on LD and individual relatedness

### [S_line_genotype_correction.R](https://github.com/tshalev/WRC-genome-paper/blob/main/Scripts/S_line_genotype_correction.R)
Manual correction of SNP data for complete selfing lines. Further fine-scale manual correction was carried out by hand.

### [S_line_pop_gen_and_heterozygosity.R](https://github.com/tshalev/WRC-genome-paper/blob/main/Scripts/S_line_pop_gen_and_heterozygosity.R)
Analysis of heterozygosity, statistical tests, chi-squared test for analysis of loci under selection during selfing (outlier SNPs), plots used in Figure 4

### [inbreeding_coeffcients.R](https://github.com/tshalev/WRC-genome-paper/blob/main/Scripts/inbreeding_coefficients.R)
Analysis and plots for inbreeding coefficients in Figure 4

### [variant_effect_predictions_and_snp_annotation.R](https://github.com/tshalev/WRC-genome-paper/blob/main/Scripts/variant_effect_predictions_and_snp_annotation.R)
Analysis of outlier SNPs during selfing, variant effects, statistical analyses of SNP consequences and GO categories, annotation of SNPs, plots used in Figure 5
