# Scripts
## [geographic_maps.R](https://github.com/tshalev/WRC-genome-paper/tree/main/Scripts/geographic_maps)
Geographic map used in Figure 1

## [genetic_differentiation.R](https://github.com/tshalev/WRC-genome-paper/main/Scripts/genetic_differentiation)
Fst analysis, DAPC, PCA plots used in Figure 1

## [LD_plots.R](https://github.com/tshalev/WRC-genome-paper/tree/main/Scripts/LD_plots)
Generating LD plots used in Figure 1. Code derived and adapted from a post made here: https://www.biostars.org/p/347796/

## [nucleotide_diversity_analysis.R](https://github.com/tshalev/WRC-genome-paper/main/Scripts/nucleotide_diversity_analysis)
Analysis of pi, dxy in pixy, plots used in Figure 3, statistical tests

## [S_line_genotype_correction.R](https://github.com/tshalev/WRC-genome-paper/main/Scripts/S_line_genotype_correction)
Manual correction of SNP data for complete selfing lines. Further fine-scale manual correction was carried out by hand.

## [S_line_pop_gen_and_heterozygosity.R](https://github.com/tshalev/WRC-genome-paper/tree/main/Scripts/S_line_pop_gen_and_heterozygosity)
Analysis of heterozygosity, statistical tests, chi-squared test for analysis of loci under selection during selfing (outlier SNPs), plots used in Figure 4

## [inbreeding_coeffcients.R](https://github.com/tshalev/WRC-genome-paper/main/Scripts/inbreeding_coefficients)
Analysis and plots for inbreeding coefficients in Figure 4

## [variant_effect_predictions_and_snp_annotation.R](https://github.com/tshalev/WRC-genome-paper/main/Scripts/variant_effect_predictions_and_snp_annotation)
Analysis of outlier SNPs during selfing, variant effects, statistical analyses of SNP consequences and GO categories, annotation of SNPs, plots used in Figure 5
